### ISSUE TITLE
Use a GCP authentication library instead of a .json file

### ISSUE THREAD

**Issue by pseguin2011** [ThouCheese/cloud-storage-rs#69]
* It would be great to remove the dependency of the  `GOOGLE_CLOUD_CREDENTIALS` or `SERVICE_ACCOUNT` environment variables. As they are not provided in Google CloudBuild or CloudRun environments.
* I would recommend using a library such as [go-auth](https://crates.io/crates/goauth) or [yup-oauth2](https://crates.io/crates/yup-oauth2) where we could use the tokens to authenticate.

**Comment by nlfiedler** [comment]
I second the vote for not using environment variables, or at least providing a simple API for providing my own credentials. In an application that takes configuration from a user, having a file on disk and setting an environment variable in an already running application is less than ideal.

**Comment by ThouCheese** [comment]
I agree that this would be a nice improvement, and I think I can cook up a nice API once #65 resolves. I imagine something along the lines of

```rust
let client = cloud_storage::Client::builder()
   .credentials(<whatever mechanism we choose>)
   .build()?;
```

Would that be sufficient for your use case?

**Comment by nlfiedler** [comment]
That would be splendid, thanks.

**Comment by pseguin2011** [comment]
yeah that sounds good, I'd actually like to take this on myself!

**Comment by ThouCheese** [comment]
Awesome! Should I ping you once the Client PR lands?

**Comment by pseguin2011** [comment]
That would be great!

**Comment by nhynes** [comment]
Alternatively, have an [extensible trait](https://docs.rs/cloud-storage-lite/latest/cloud_storage_lite/token_provider/trait.TokenProvider.html) that allows people to bring their own providers (e.g., [secret injection](https://github.com/banzaicloud/bank-vaults))

**Comment by ThouCheese** [comment]
@pseguin2011 ping!

**Comment by pseguin2011** [comment]
Alright, I'll get started on this :+1:

**Comment by pseguin2011** [comment]
Sorry It took so long! here you go!

**Comment by hendrikmaus** [comment]
Another vote for using a shared auth lib as it increases the chance to support Google's Workload Identity feature.

**Comment by GlenDC** [comment]
Any update on this @ThouCheese , that PR has been open forever without much interaction. Is there any help required that people can offer? It's a shame that this otherwise great library doesn't seem to be flexible in the kind of credentials it receives. E.g. the assumption that it's always a service account doesn't make it useable for all my use cases.

**Comment by ThouCheese** [comment]
Hi @GlenDC and @hendrikmaus and @pseguin2011! Sorry that I have not been as responsive as I should have been, I have not been able to devote as much time due to personal reasons. For now:

This PR looks good enough so I will merge it. I believe that since we are giving a default value to the generic client that preserves the old behaviour, this is not a breaking change. Therefore I will release it as `0.11.3` and we can experiment with whether this is a sufficient approach.

**Comment by GlenDC** [comment]
Fantastic news! Going to play around with the merged MR. Very exciting, were there any issues you think of that still need to be tackled or could be improved upon @pseguin2011 ?

**Comment by ThouCheese** [comment]
I want to experiment a bit with maybe adding a default to the type parameter of `Client` to ensure that this change is not that painful for the users; people that use custom TokenCache implementation can just specify then, people that want the old behaviour can have it just work. If you wanna do some experimentation that is great! I will probably release this to crates.io together with some other PR's, so getting some feedback in soon would be nice :)
